### DISCUSSION TITLE
Logger's sinks accessor

### DISCUSSION THREAD

**Discussion by jol** [odygrd/quill#593] [Q&A]
Hi,

I would like to know if it will be possible to add a kind of:
```cpp

auto& LoggerBase::get_sinks()
{
    return sinks;
}
```

In the previous version of quill, If I remember correctly, we were able to create a root logger with multiple handlers, then when another logger was created, it "inherits" the handlers from the root logger.

I would like to be able to do something equivalent, like:
```cpp
auto root_logger  = quill::Frontend::create_or_get_logger("root", {sink1, sink2});
...

// then maybe in another class / thread ... 
auto my_logger = quill::Frontend::create_or_get_logger("root", quill::Frontend::get_logger("root")->get_sinks());

```

What do you think about this point, maybe there is a better way to achieve this.

Regards,

Jol

**Comment by odygrd** [comment]
Hey,

Yeah, that was removed in the latest versions because it was adding some extra complexity, especially with the default logger, and I didn’t find it particularly useful since loggers are usually created at the start of the application anyway.

Adding an accessor to the logger base feels a bit risky to expose in the API. The issue is that the sinks are managed by the backend worker thread, and if we expose the pointer, there's a chance someone might accidentally call a function on it, which could lead to problems. Although, to be fair, the sinks are already accessible in a similar way if you use `Frontend::get_sink(...)` anyway...

What we could do instead is add a `create_or_get_logger ` overload to accept a `Logger*` and do this internally, so we get the same result without exposing anything to the user. I’ll take a look and see if that could work! 👍

**Reply by jol** [reply]
Thank you for your reply.

About the Frontend::get_sink(...) it means I need to know the name & all sinks attached to my root logger. 

your suggestion of accepting a `Logger*` to the `create_or_get_logger` will be perfect!

**Reply by odygrd** [reply]
> About the Frontend::get_sink(...) it means I need to know the name & all sinks attached to my root logger.

yeah i meant that even without the logger sink accessor we don’t currently have total safety

**Reply by odygrd** [reply]
done in https://github.com/odygrd/quill/pull/597 :)

**Comment by jol** [comment]
In the same idea, I would have access from the Logger to the user_clock & and clock_source. 

I'm in a situation where I have a root logger with some specific sinks and a custom user_clock. With the previous development, I'm able to have other Loggers that " inherit" from the root_logger. Now, for some loggers, I would have some dedicated sink but still use the same user_clock and clock_source as the root logger.

Having public accessors would be excellent for getting user_clock & clock_source.

Jol

**Reply by odygrd** [reply]
hey, this is done in https://github.com/odygrd/quill/pull/600
Can you use the master branch, or do you need another release?

**Comment by jol** [comment]
Perfect, I can use the master branch !

Thank you
